Atangster/314 Add solidpod login theme configuration

atangster opened the pull request:
## Pull Request Details

Link to associated issue: https://github.com/anusii/healthpod/issues/314, https://github.com/anusii/healthpod/issues/160

Related PRs from Healthpod project:
- https://github.com/anusii/healthpod/pull/327
- https://github.com/anusii/healthpod/pull/169 (no longer relevant)

# Checklist
Complete the check-list below to ensure your branch is ready for PR.

Flutter Style Guide: https://survivor.togaware.com/gnulinux/flutter-style.html

- [x] Screenshots included in linked issue
- [x] Changes adhere to the team style and coding guideline
- [x] No confidential information
- [x] No duplicated content
- [x] No lint check errors related to your changes (make prep or flutter analyze lib)
- [x] Tested on at least one device
  - [x] Windows
  - [x] Linux 
- [x] Added 2 reviewers (or 1 for private repositories then they add another)

gjwgit reviewed (approved):
Works well for me in healthpod - both dark mode and login (with a couple of idiosyncrasies that may be independent of this PR).

@anushkavidanage and @cdawei can you please check that this PR does not affect your apps. Thanks.

Once we have the three approvals and it is merged I will submit a new release to pub.dev

cdawei commented:
I tested the example app in this pull request (i.e., `solidpod/example`) and found a few issues:
- Clicking the `Login` button launches a browser instead of the in-app webview (is this one of the purpose of this PR?)
- The server URL text is black which is unreadable
- Missing flutter platform files including `pubspec.yaml` and image assets (this is not critical but would be better to have them ready as an example app)

Perhaps porting a few settings from healthpod to `example/lib/main.dart` could resolve the first 2 issues?

Here's a workaround for the 3rd issue before I can run the example app.
- creating linux platform files: `flutter create --platforms=linux .`
- add `solidpod` to `pubspec.yaml`: `flutter pub add 'solidpod:{"path":"../"}'`
- add image assets and set the `image` and `logo` arguments of `SolidLogin` in `example/lib/main.dart`
- add override dependencies (for linux app) as per [this comment](https://github.com/juliansteenbakker/flutter_secure_storage/issues/829#issuecomment-2766270627).
```
dependency_overrides:
  flutter_secure_storage_linux: ^2.0.1
  flutter_secure_storage_platform_interface: ^2.0.0
```

atangster commented:
@cdawei thanks for the feedback!

Made some updates:
- Setup the example project including configuring linux app, added platform files
- Fixed server text readability
- Login button should by default take user to browser (removed the old inappwebview auth code as not compatible with linux platforms and we have found an alternative using chromedriver headless browser for healthpod project, see https://github.com/anusii/healthpod/issues/208 and https://github.com/anusii/healthpod/issues/189)

Example project login screen:

![image](https://github.com/user-attachments/assets/5615c353-3c65-41d1-aa11-f0f54826eff2)

cdawei reviewed (approved):
Worked as expected on Ubuntu 24.10, thanks @atangster.

anushkavidanage commented:
Tested this with `exampleapp`, `demopod`, and `riopod`. I've got a few observations/questions.

1. It seems like the dark mode is built in. Why don't we let the user decide which mode (dark or light) they need, and define an optional parameter in `SolidLogin` function to set this up with a default mode?
2. Why do we have two example apps within the `solidpod`, `exampleapp` and `demopod`? I think we only need one with all the example functionalities in it. I suggest to remove `example app` and rename the `demopod` as `example`.
3. When I built all three of the above apps using the new `solidpod` in Windows I get the following warning. The app gets built but with these warnings. I wonder if that is specific to Windows?

```
<path>\riopod\windows\flutter\ephemeral\.plugin_symlinks\flutter_inappwebview_windows\windows\types\web_resource_response.cpp(54,28): warning C4244: 'argument': conversion from '__int64' to 'int', possible loss of data [<path>\riopod\build\windows\x64\plugins\flutter_inappwebview_windows\flutter_inappwebview_windows_plugin.vcxproj]
<path>\riopod\build\windows\x64\packages\Microsoft.Web.WebView2\build\native\include\WebView2EnvironmentOptions.h(194,3): warning C4458: declaration of 'value' hides class member [<path>\riopod\build\windows\x64\plugins\flutter_inappwebview_windows\flutter_inappwebview_windows_plugin.vcxproj]
<path>\riopod\build\windows\x64\packages\Microsoft.Web.WebView2\build\native\include\WebView2EnvironmentOptions.h(193,3): warning C4458: declaration of 'value' hides class member [<path>\riopod\build\windows\x64\plugins\flutter_inappwebview_windows\flutter_inappwebview_windows_plugin.vcxproj]
<path>\riopod\build\windows\x64\packages\Microsoft.Web.WebView2\build\native\include\WebView2EnvironmentOptions.h(194,3): warning C4458: declaration of 'value' hides class member [<path>\riopod\build\windows\x64\plugins\flutter_inappwebview_windows\flutter_inappwebview_windows_plugin.vcxproj]
<path>\riopod\build\windows\x64\packages\Microsoft.Web.WebView2\build\native\include\WebView2EnvironmentOptions.h(193,3): warning C4458: declaration of 'value' hides class member [<path>\riopod\build\windows\x64\plugins\flutter_inappwebview_windows\flutter_inappwebview_windows_plugin.vcxproj]
```

gjwgit commented:
@anushkavidanage maybe try clearing the cache to rid the warnings?

atangster commented:
Added theme light/dark mode toggle to be consistent with healthpod implementation:

https://github.com/user-attachments/assets/48520a4d-e645-42fb-8457-4339bbee256e

Renamed `demopod` to `example` project

@anushkavidanage thanks for your feedback! Ready for re-review

atangster committed:
add theme toggle and use demopod as example

atangster committed:
format cmt

gjwgit commented:
Once reviewed by @anushkavidanage then @gjwgit will test, review, and publish.

anushkavidanage reviewed (approved):
Tested. All working well. 

One last thing to check is whether we need all the build files in this PR. I see a large number of files related to specific OS versions. I'm not sure we need all those for this PR. Please double-check and remove unnecessary files before merging.

gjwgit commented:
Thanks @anushkavidanage.

@anushkavidanage can you remove anything from git not required in the example folder. In markdown_tooltip, for example, I just have the following:

![image](https://github.com/user-attachments/assets/894c42f8-b632-4e1b-bbaf-81e9d88a26a3)

Many of the other files will be generated by flutter as needed by a user compiling the app, so they can be removed from git.

Thanks

atangster committed:
remove unnecesary build files

atangster commented:
- [x] Remove build files from `demopod` (now named `example`) 
- [x] Update Makefile

Updated `example` structure:

![image](https://github.com/user-attachments/assets/51debeb7-c151-4346-adab-739e3c3fccf2)

@gjwgit ready for re-review

gjwgit commented:
In @atangster fork. Do we need to update the flutter_secure_storage in `solidpod`? Currently the anusii repo demopod works okay.

```
13 Apr 14:37:12 gjw@kadesh ~ solidpod/example$ flutter run -d linux
Launching lib/main.dart on Linux in debug mode...
Building Linux application...                                   
/home/gjw/Remove/solidpod/example/linux/flutter/ephemeral/.plugin_symlinks/flutter_secure_storage_linux/linux/flutter_secure_storage_linux_plugin.cc:193:50: error: use of undeclared identifier 'APPLICATION_ID'
/home/gjw/Remove/solidpod/example/linux/flutter/ephemeral/.plugin_symlinks/flutter_secure_storage_linux/linux/flutter_secure_storage_linux_plugin.cc:194:67: error: use of undeclared identifier 'APPLICATION_ID'
Error: Build process failed

...

13 Apr 14:37:57 gjw@kadesh ~ solidpod/example$ cat ../.git/config 
[core]
	repositoryformatversion = 0
	filemode = true
	bare = false
	logallrefupdates = true
[remote "origin"]
	url = git@github.com:atangster/solidpod.git
	fetch = +refs/heads/*:refs/remotes/origin/*
[branch "dev"]
	remote = origin
	merge = refs/heads/dev

```

atangster committed:
fix flutter_secure_storage linux error

atangster commented:
The above Linux build error appears to be related to this: https://github.com/juliansteenbakker/flutter_secure_storage/issues/829

The issue occurs because `flutter_secure_storage 9.2.4` depends on `flutter_secure_storage_linux 1.2.2`, which doesn't properly define APPLICATION_ID in its Linux plugin. The fix is actually in `flutter_secure_storage_linux 2.0.1`.

- [x] Add `flutter_secure_storage_linux ^2.0.1` as a dependency, which overrides the transitive dependency with the fixed version

gjwgit reviewed (commented):
```
14 Apr 05:31:04 gjw@kadesh ~ solidpod/example$ dart pub get
Resolving dependencies... 
Because no versions of flutter_secure_storage match >9.2.4 <10.0.0 and flutter_secure_storage 9.2.4 depends on flutter_secure_storage_linux ^1.2.2,
  flutter_secure_storage ^9.2.4 requires flutter_secure_storage_linux ^1.2.2.
And because every version of solidpod from path depends on flutter_secure_storage ^9.2.4, every version of solidpod from path requires
  flutter_secure_storage_linux ^1.2.2.
So, because demopod depends on solidpod from path which depends on flutter_secure_storage_linux ^2.0.1, version solving failed.
```

cdawei commented:
> ```
> 14 Apr 05:31:04 gjw@kadesh ~ solidpod/example$ dart pub get
> Resolving dependencies... 
> Because no versions of flutter_secure_storage match >9.2.4 <10.0.0 and flutter_secure_storage 9.2.4 depends on flutter_secure_storage_linux ^1.2.2,
>   flutter_secure_storage ^9.2.4 requires flutter_secure_storage_linux ^1.2.2.
> And because every version of solidpod from path depends on flutter_secure_storage ^9.2.4, every version of solidpod from path requires
>   flutter_secure_storage_linux ^1.2.2.
> So, because demopod depends on solidpod from path which depends on flutter_secure_storage_linux ^2.0.1, version solving failed.
> ```

Does the dependency override mentioned [here](https://github.com/anusii/solidpod/pull/267#issuecomment-2781945634) help?

atangster committed:
add dependency overrides

atangster commented:
@gjwgit I added dependency overrides as suggested by @cdawei 

Dependency issues should be resolved now, feel free to test

gjwgit commented:
Looks like it is an active issue for flutter_secure_storage and hopefully to be fixed soon.

@dawei/@atangster  do you know if there is some way to add these overrides in the package pubspec rather than in the app pubspec so we don't break current apps relying on solidpod?

Otherwise do we delay until the issue is fixed?

cdawei commented:
> @dawei/@atangster do you know if there is some way to add these overrides in the package pubspec rather than in the app pubspec so we don't break current apps relying on solidpod?

Noted the following in Dart official documentation [here](https://dart.dev/tools/pub/dependencies#dependency-overrides).

---
*Only the dependency overrides in a package's own pubspec are considered during package resolution. Dependency overrides inside any depended-on packages are ignored.*

*As a result, if you publish a package to pub.dev, keep in mind that your package's dependency overrides are ignored by all users of your package.*

atangster committed:
Merge remote-tracking branch 'upstream/dev' into dev

atangster commented:
Resolved conflict, ready for re-review/merging 
CC: @gjwgit

gjwgit commented:
We will merge into anusii dev branch and continue testing there. The overrides from the app side remain necessary for now and as @cdawei notes this is not possible at the package level.
